File: go-sdk/starlight/messages.go
package starlight

import (
	"encoding/json"
	"fmt"
	"time"
)

// Message represents a JSON-RPC 2.0 message used in the Starlight Protocol.
type Message struct {
	JSONRPC string          `json:"jsonrpc"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	ID      string          `json:"id,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *RPCError       `json:"error,omitempty"`
}

// RPCError represents a JSON-RPC 2.0 error.
type RPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitempty"`
}

// RegistrationParams contains parameters for sentinel registration.
type RegistrationParams struct {
	Layer        string   `json:"layer"`
	Priority     int      `json:"priority"`
	Capabilities []string `json:"capabilities,omitempty"`
	Selectors    []string `json:"selectors,omitempty"`
	AuthToken    string   `json:"authToken,omitempty"`
}

// PreCheckParams contains parameters sent by Hub during pre-check.
type PreCheckParams struct {
	Command    CommandInfo      `json:"command"`
	Blocking   []BlockingElement `json:"blocking,omitempty"`
	Screenshot string           `json:"screenshot,omitempty"`
	URL        string           `json:"url,omitempty"`
}

// CommandInfo describes the pending command.
type CommandInfo struct {
	Cmd           string `json:"cmd,omitempty"`
	Goal          string `json:"goal,omitempty"`
	Selector      string `json:"selector,omitempty"`
	Value         string `json:"value,omitempty"`
	StabilityHint int    `json:"stabilityHint,omitempty"`
}

// BlockingElement describes an element that might block execution.
type BlockingElement struct {
	Selector string `json:"selector"`
	Tag      string `json:"tag,omitempty"`
	Classes  string `json:"classes,omitempty"`
	Text     string `json:"text,omitempty"`
}

// EntropyStreamParams contains entropy data from the Hub.
type EntropyStreamParams struct {
	Entropy        bool `json:"entropy"`
	MutationCount  int  `json:"mutationCount,omitempty"`
	NetworkPending int  `json:"networkPending,omitempty"`
}

// ActionParams contains parameters for sentinel actions during hijack.
type ActionParams struct {
	Cmd      string `json:"cmd"`
	Selector string `json:"selector"`
	Text     string `json:"text,omitempty"`
}

// ContextUpdateParams contains sovereign context data.
type ContextUpdateParams struct {
	Context map[string]any `json:"context"`
}

// WaitParams contains parameters for the wait response.
type WaitParams struct {
	RetryAfterMs int `json:"retryAfterMs,omitempty"`
}

// HijackParams contains parameters for hijack requests.
type HijackParams struct {
	Reason string `json:"reason"`
}

// ResumeParams contains parameters for resume after hijack.
type ResumeParams struct {
	ReCheck bool `json:"re_check"`
}

// NewMessage creates a new JSON-RPC 2.0 message.
func NewMessage(method string, params any) (*Message, error) {
	var rawParams json.RawMessage
	if params != nil {
		data, err := json.Marshal(params)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal params: %w", err)
		}
		rawParams = data
	}

	return &Message{
		JSONRPC: "2.0",
		Method:  method,
		Params:  rawParams,
		ID:      fmt.Sprintf("%d", time.Now().UnixNano()),
	}, nil
}

// NewResponse creates a response message with the same ID.
func NewResponse(id string, method string, params any) (*Message, error) {
	msg, err := NewMessage(method, params)
	if err != nil {
		return nil, err
	}
	msg.ID = id
	return msg, nil
}

// ParseParams unmarshals the params into the provided struct.
func (m *Message) ParseParams(v any) error {
	if m.Params == nil {
		return nil
	}
	return json.Unmarshal(m.Params, v)
}

// Marshal serializes the message to JSON bytes.
func (m *Message) Marshal() ([]byte, error) {
	return json.Marshal(m)
}

// ParseMessage deserializes JSON bytes into a Message.
func ParseMessage(data []byte) (*Message, error) {
	var msg Message
	if err := json.Unmarshal(data, &msg); err != nil {
		return nil, fmt.Errorf("failed to parse message: %w", err)
	}
	return &msg, nil
}

File: rust-sdk/src/messages.rs
//! JSON-RPC 2.0 and Starlight Protocol message types.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// JSON-RPC 2.0 version constant.
pub const JSONRPC_VERSION: &str = "2.0";

// =============================================================================
// Base JSON-RPC Types
// =============================================================================

/// A JSON-RPC 2.0 request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcRequest<T> {
    pub jsonrpc: String,
    pub method: String,
    pub params: T,
    pub id: String,
}

impl<T> JsonRpcRequest<T> {
    /// Create a new JSON-RPC request.
    pub fn new(method: impl Into<String>, params: T, id: impl Into<String>) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            method: method.into(),
            params,
            id: id.into(),
        }
    }
}

/// A JSON-RPC 2.0 response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcResponse<T> {
    pub jsonrpc: String,
    pub result: Option<T>,
    pub error: Option<JsonRpcError>,
    pub id: String,
}

/// A JSON-RPC 2.0 error.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

/// A JSON-RPC 2.0 notification (no id field).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcNotification<T> {
    pub jsonrpc: String,
    pub method: String,
    pub params: T,
}

impl<T> JsonRpcNotification<T> {
    /// Create a new notification.
    pub fn new(method: impl Into<String>, params: T) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            method: method.into(),
            params,
        }
    }
}

// =============================================================================
// Starlight Protocol Message Types
// =============================================================================

/// Registration parameters for Sentinel → Hub.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegistrationParams {
    /// Sentinel layer name (e.g., "JanitorSentinel")
    pub layer: String,

    /// Priority (1-10, lower = higher priority)
    pub priority: u8,

    /// Capabilities (e.g., ["detection", "healing"])
    #[serde(default)]
    pub capabilities: Vec<String>,

    /// CSS selectors this Sentinel monitors
    #[serde(default)]
    pub selectors: Vec<String>,

    /// Optional JWT authentication token
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth_token: Option<String>,
}

impl RegistrationParams {
    /// Create new registration params.
    pub fn new(layer: impl Into<String>, priority: u8) -> Self {
        Self {
            layer: layer.into(),
            priority,
            capabilities: Vec::new(),
            selectors: Vec::new(),
            auth_token: None,
        }
    }

    /// Add capabilities.
    pub fn with_capabilities(mut self, caps: Vec<impl Into<String>>) -> Self {
        self.capabilities = caps.into_iter().map(Into::into).collect();
        self
    }

    /// Add selectors.
    pub fn with_selectors(mut self, sels: Vec<impl Into<String>>) -> Self {
        self.selectors = sels.into_iter().map(Into::into).collect();
        self
    }

    /// Set auth token.
    pub fn with_auth_token(mut self, token: impl Into<String>) -> Self {
        self.auth_token = Some(token.into());
        self
    }
}

/// Pre-check parameters from Hub → Sentinel.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreCheckParams {
    /// Current page URL
    #[serde(default)]
    pub url: Option<String>,

    /// Upcoming command type
    pub command: String,

    /// Target selector for the command
    #[serde(default)]
    pub selector: Option<String>,

    /// Semantic goal
    #[serde(default)]
    pub goal: Option<String>,

    /// Detected blocking elements
    #[serde(default)]
    pub blocking: Vec<BlockingElement>,

    /// Page screenshot (base64)
    #[serde(default)]
    pub screenshot: Option<String>,

    /// Additional context
    #[serde(default)]
    pub context: HashMap<String, serde_json::Value>,
}

/// A blocking element detected by the Hub.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockingElement {
    pub selector: String,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub element_type: Option<String>,
}

/// Pre-check response types.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "response")]
pub enum PreCheckResponse {
    /// All clear - proceed with command
    #[serde(rename = "clear")]
    Clear,

    /// Wait - retry after specified milliseconds
    #[serde(rename = "wait")]
    Wait {
        #[serde(rename = "retryAfterMs")]
        retry_after_ms: u64,
        #[serde(default)]
        reason: Option<String>,
    },

    /// Hijack - Sentinel takes browser control
    #[serde(rename = "hijack")]
    Hijack { reason: String },
}

/// Hijack parameters (Sentinel → Hub).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HijackParams {
    pub reason: String,
}

/// Action command during hijack (Sentinel → Hub).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionParams {
    pub cmd: ActionCommand,
    pub selector: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
}

/// Available action commands during hijack.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ActionCommand {
    Click,
    Fill,
    Hide,
    Remove,
}

/// Resume parameters after hijack.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeParams {
    /// Request re-check after resume
    #[serde(default = "default_true")]
    pub request_recheck: bool,
}

fn default_true() -> bool {
    true
}

/// Context update from Sentinel.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextUpdateParams {
    pub context: HashMap<String, serde_json::Value>,
}

/// Entropy (page state) update from Hub.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntropyParams {
    pub url: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub mutations: u32,
    #[serde(default)]
    pub network_pending: u32,
    #[serde(default)]
    pub context: HashMap<String, serde_json::Value>,
}

// =============================================================================
// Protocol Method Names
// =============================================================================

/// Starlight Protocol method names.
pub mod methods {
    pub const REGISTRATION: &str = "starlight.registration";
    pub const PRE_CHECK: &str = "starlight.pre_check";
    pub const CLEAR: &str = "starlight.clear";
    pub const WAIT: &str = "starlight.wait";
    pub const HIJACK: &str = "starlight.hijack";
    pub const ACTION: &str = "starlight.action";
    pub const RESUME: &str = "starlight.resume";
    pub const ENTROPY: &str = "starlight.entropy";
    pub const CONTEXT_UPDATE: &str = "starlight.context_update";
    pub const INTENT: &str = "starlight.intent";
}

// =============================================================================
// Helper Types
// =============================================================================

/// Raw incoming message that can be either a request or notification.
#[derive(Debug, Clone, Deserialize)]
pub struct RawMessage {
    pub jsonrpc: String,
    pub method: String,
    pub params: serde_json::Value,
    pub id: Option<String>,
}
